otx/algorithms/segmentation/adapters/mmseg/models/segmentors/mean_teacher_segmentor.py

```diff
@@ -42,6 +67,49 @@ def encode_decode(self, img, img_metas):
         """Encode and decode images."""
         return self.model_s.encode_decode(img, img_metas)
 
+    def decode_proto_network(
+        self, sup_input, gt_semantic_seg, unsup_input=None, pl_from_teacher=None, reweight_unsup=1.0
+    ):
+        """Forward prototype network, compute proto loss."""
+        # supervised branch
+        head_features_sup = self.model_s.decode_head._forward_feature(sup_input)
+        proto_out_supervised = self.proto_net(head_features_sup, gt_semantic_seg)
+        loss_proto = self.proto_net.losses(**proto_out_supervised, seg_label=gt_semantic_seg)
+        self._update_summary_loss(loss_proto, loss_weight=self.proto_weight)
+        # unsupervised branch
+        if unsup_input is not None and pl_from_teacher is not None:
+            head_features_unsup = self.model_s.decode_head._forward_feature(unsup_input)
+            proto_out_unsupervised = self.proto_net(head_features_unsup, pl_from_teacher)
+            loss_proto_u = self.proto_net.losses(**proto_out_unsupervised, seg_label=pl_from_teacher)
+            self._update_summary_loss(loss_proto_u, loss_weight=self.unsup_weight * reweight_unsup * self.proto_weight)
+
+    def generate_pseudo_labels(self, ul_w_img, ul_img_metas):
+        """Generate pseudo labels from teacher model, apply filter loss method."""
+        with torch.no_grad():
+            teacher_feat = self.model_t.extract_feat(ul_w_img)
+            teacher_out = self.model_t._decode_head_forward_test(teacher_feat, ul_img_metas)
+            teacher_out = resize(
+                input=teacher_out, size=ul_w_img.shape[2:], mode="bilinear", align_corners=self.align_corners
+            )
+            teacher_prob_unsup = torch.softmax(teacher_out, axis=1)
+            _, pl_from_teacher = torch.max(teacher_prob_unsup, axis=1, keepdim=True)
+
+        # drop pixels with high entropy
```

Comment:
I have a question about dropping pixels with high entropy. Is this from original prototype based paper? Does it drop all pixel predictions masking it as ignored label 255 rather than using threshold 0.5 for prediction logit? Drop percent 80 looks so high, which the prediction per label is all dismissed.

Reply:
It is actually from U2PL paper, as I remember. I utilized this method since it works pretty well. The naming of drop pixels = 80% is misleading. I will rename it. Actually 20% of pixels are droped from start of training and this number decreases during training.

`percent_unreliable = (100 - drop_percent) * (1 - self.count_iter / self.filter_pixels_iters)
drop_percent = 100 if percent_unreliable <= 0 else 100 - percent_unreliable
 thresh = np.percentile(entropy[pl_from_teacher != 255].detach().cpu().numpy().flatten(), drop_percent)
 thresh_mask = entropy.ge(thresh).bool() * (pl_from_teacher != 255).bool()`
 
 So in first epoch we drop 20% (we compute 0.8 percentile and drop everything greater). In the following epochs with percentage decreases.